src/stdlib_linalg.fypp

```diff
@@ -108,4 +201,200 @@ contains
         end do
       end function trace_${t1[0]}$${k1}$
     #:endfor
+
+
+    #:for k1, t1 in RCI_KINDS_TYPES
+      pure function is_square_${t1[0]}$${k1}$(A) result(res)
+        ${t1}$, intent(in) :: A(:,:)
+        logical :: res
+        res = (size(A,1) .eq. size(A,2))
+      end function is_square_${t1[0]}$${k1}$
+    #:endfor
+
+
+    #:for k1, t1 in RCI_KINDS_TYPES
+      pure function is_diagonal_${t1[0]}$${k1}$(A) result(res)
+        ${t1}$, intent(in) :: A(:,:)
+        logical :: res
+        ${t1}$ :: zero
+        integer :: m, n, o, i, j
+        zero = 0 !zero of relevant type
+        m = size(A,1)
+        n = size(A,2)
+        do j = 1, n !loop over all columns
+            o = min(j-1,m) !index of row above diagonal (or last row)
+            do i = 1, o !loop over rows above diagonal
+                if (.not. (A(i,j) .eq. zero)) then
+                  res = .false.
+                  return
+                end if
+            end do
+            do i = o+2, m !loop over rows below diagonal
+                if (.not. (A(i,j) .eq. zero)) then
```

Comment:
```suggestion
                if ((A(i,j) .ne. zero)) then
```

Is their some reason you prefer the explicit negation instead of using the *not equal* (`.ne.` or `/=`) operator? This also applies in other places.

I guess a good compiler will probably rewrite this to minimize operations anyways, but it's still good to clear up what the intentions are.

Reply:
No good reason. I always forget which language use which negated equality with symbols (between `!=, ~=, /=`, and more), so I just reverted to something I knew would work (while forgetting about `.neq.`).

I'll make the changes, thanks for catching this.

Reply:
All changes should be made. Thanks again.